core_lib/src/managers/selectionmanager.cpp

```diff
@@ -0,0 +1,370 @@
+
+//#ifdef QT_DEBUG
+#include <QDebug>
+//#endif
+
+
+SelectionManager::SelectionManager(Editor* editor) : BaseManager(editor),
+    mySelection(QRectF()), myTempTransformedSelection(QRectF()), myTransformedSelection(QRectF()),
+    mSomethingSelected(false), mLastSelectionPolygonF(QPolygonF()), mCurrentSelectionPolygonF(QPolygonF()),
+    myRotatedAngle(0), mOffset(QPointF()), mMoveMode(MoveMode::NONE)
+{
+}
+
+SelectionManager::~SelectionManager()
+{
+}
+
+bool SelectionManager::init()
+{
+    return true;
+}
+
+Status SelectionManager::load(Object*)
+{
+    return Status::OK;
+}
+
+Status SelectionManager::save(Object*)
+{
+    return Status::OK;
+}
+
+void SelectionManager::workingLayerChanged(Layer *)
+{
+}
+
+void SelectionManager::resetSelectionTransformProperties()
+{
+    mOffset = QPoint(0, 0);
+    myRotatedAngle = 0;
+    mSelectionTransform.reset();
+}
+
+
+void SelectionManager::updatePolygons()
+{
+    mCurrentSelectionPolygonF = myTempTransformedSelection;
+    mLastSelectionPolygonF = myTransformedSelection;
+}
+
+void SelectionManager::resetSelectionTransform()
+{
+    mSelectionTransform.reset();
+}
+
+bool SelectionManager::isOutsideSelectionArea(QPointF point)
+{
+    return (!myTransformedSelection.contains(point)
+            && validateMoveMode(point) == MoveMode::NONE);
+}
+
+void SelectionManager::deleteSelection()
+{
+    emit needDeleteSelection();
+}
+
+void SelectionManager::clearCurves()
+{
+    mClosestCurves.clear();
+}
+
+void SelectionManager::clearVertices()
+{
+    mClosestVertices.clear();
+}
+
+float SelectionManager::selectionTolerance() const
+{
+    return mSelectionTolerance * editor()->viewScaleInversed();
+}
+
+MoveMode SelectionManager::validateMoveMode(QPointF pos)
+{
+    return moveModeForAnchorInRange(pos);
+}
+
+MoveMode SelectionManager::moveModeForAnchorInRange(QPointF lastPos)
+{
+    const double marginInPixels = 15;
+    const double scale = editor()->viewScaleInversed();
+    const double scaledMargin = qAbs(marginInPixels * scale);
+
+    qDebug() << editor()->viewScaleInversed();
+
+    QRectF transformRect = myTempTransformedSelection;
+    QPointF lastPoint = lastPos;
+
+    MoveMode mode;
+    if (QLineF(lastPoint, transformRect.topLeft()).length() < scaledMargin)
+    {
+        mode = MoveMode::TOPLEFT;
+    }
+    else if (QLineF(lastPoint, transformRect.topRight()).length() < scaledMargin)
+    {
+        mode = MoveMode::TOPRIGHT;
+    }
+    else if (QLineF(lastPoint, transformRect.bottomLeft()).length() < scaledMargin)
+    {
+        mode = MoveMode::BOTTOMLEFT;
+
+    }
+    else if (QLineF(lastPoint, transformRect.bottomRight()).length() < scaledMargin)
+    {
+        mode = MoveMode::BOTTOMRIGHT;
+    }
+    else if (myTransformedSelection.contains(lastPoint))
+    {
+        mode = MoveMode::MIDDLE;
+    }
+    else {
+        mode = MoveMode::NONE;
+    }
+    mMoveMode = mode;
+    return mode;
+}
+
+MoveMode SelectionManager::getMoveModeForSelectionAnchor(QPointF pos)
```

Comment:
Yes, because the range has to start from an offset from center and not center itself, otherwise you'll never be in range. Visually the difference is:
The left is what it would look like if we used the mCurrentSelectionPolygonF[x], the right is the current implementation.
<img width="734" alt="image" src="https://user-images.githubusercontent.com/1045397/59833636-c5f9ff80-9346-11e9-8d20-24dee1d7bb84.png">

Reply:
I'm still not sold on this. Say that the top left point of the selection is at `k,k` = `mCurrentSelectionPolygonF[0]`. Then `topLeftCorner` is a rectangle from `k-radius, k-radius` to `k+radius, k+radius`. Then in the if statements we make a line from `currentPos` to the center of `topLeftCorner`. The center of `topLeftCorner` is the average of the two ends: `((k+radius)+(k-radius))/2` = `k` which is the same as the initial point! I agree that the range has to be offset, but we don't need the range, we only need the center of the circle in which clicking on a corner is registered.

Reply:
I'm not following, what do you want me to change then? is it the if statements, checking for the length or the way I create the QRect.. It's getting a bit too abstract for me to understand now ?

Reply:
I am suggesting you remove the `QRectF`s entirely and replace `topLeftCorner.center()`, `topRightCorner.center()`, etc. with `mCurrentSelectionPolygonF[0]`, `mCurrentSelectionPolygonF[1]`, etc.

Reply:
Ah.... riiight I get it now :P true the margin is not needed for anything excep trying to find the center.. which is already there.